src/main/kotlin/com/manimdsl/frontend/AbstractSyntaxTree.kt

```diff
@@ -0,0 +1,73 @@
+
+// Code Specific Nodes holding line number - todo: replace with composition
+sealed class CodeNode(open val lineNumber: Int) : StatementNode()
+data class DeclarationNode(override val lineNumber: Int, val identifier: String, val expression: ExpressionNode): CodeNode(lineNumber)
+data class AssignmentNode(override val lineNumber: Int, val identifier: String, val expression: ExpressionNode): CodeNode(lineNumber)
+
+// Expressions
+sealed class ExpressionNode(override val lineNumber: Int): CodeNode(lineNumber)
+data class IdentifierNode(override val lineNumber: Int, val identifier: String): ExpressionNode(lineNumber)
+data class NumberNode(override val lineNumber: Int, val double: Double): ExpressionNode(lineNumber)
+data class MethodCallNode(override val lineNumber: Int, val instanceIdentifier: String, val dataStructureMethod: DataStructureMethod, val arguments: List<ExpressionNode>): ExpressionNode(lineNumber)
+data class ConstructorNode(override val lineNumber: Int, val type: Type, val arguments: List<ExpressionNode>): ExpressionNode(lineNumber)
+
+// Binary Expressions
+sealed class BinaryExpression(override val lineNumber: Int, open val expr1: ExpressionNode, open val expr2: ExpressionNode): ExpressionNode(lineNumber)
+data class AddExpression(override val lineNumber: Int, override val expr1: ExpressionNode, override val expr2: ExpressionNode): BinaryExpression(lineNumber, expr1, expr2)
+data class SubtractExpression(override val lineNumber: Int, override val expr1: ExpressionNode, override val expr2: ExpressionNode): BinaryExpression(lineNumber, expr1, expr2)
+data class MultiplyExpression(override val lineNumber: Int, override val expr1: ExpressionNode, override val expr2: ExpressionNode): BinaryExpression(lineNumber, expr1, expr2)
+
+// Unary Expressions
+sealed class UnaryExpression(override val lineNumber: Int, open val expr: ExpressionNode): ExpressionNode(lineNumber)
+data class PlusExpression(override val lineNumber: Int, override val expr: ExpressionNode): UnaryExpression(lineNumber, expr)
+data class MinusExpression(override val lineNumber: Int, override val expr: ExpressionNode): UnaryExpression(lineNumber, expr)
+
+
+// Types (to be used in symbol table also)
+sealed class Type: ASTNode()
+// Primitive / Data structure distinction requested by code generation
+sealed class PrimitiveType: Type()
+object NumberType: PrimitiveType()
+sealed class DataStructureType(open var internalType: Type, open val methods: HashMap<String, DataStructureMethod>): Type() {
+    abstract fun containsMethod(method: String): Boolean
+    abstract fun getMethodByName(method: String): DataStructureMethod
+}
+
+open class DataStructureMethod(open val returnType: Type, open var argumentTypes: List<Type>)
+data class ErrorMethod(override val returnType: Type = NoType, override var argumentTypes: List<Type> = listOf()) : DataStructureMethod(returnType, argumentTypes)
+
+data class StackType(override var internalType: Type = NumberType,
```

Comment:
Move all Stack stuff to separate file. This file is v cluttered. Try to make it nicer